### PR TITLE
Feature/ens asyncify

### PR DESCRIPTION
**Pull request by dbfreem**
Here is AsyncENS. Still needs docs and cleanup

There was A LOT of copy and paste here from ENS since ENS is so dependent on Eth, and Contract there wasn't much code that could be put in BaseENS.

1. There is a possible breaking change in the way the modules are loaded in ens utils.py. To be consistent between AsyncENS and ENS I only loaded the Eth module when the Web3 instance is created
2. I would like to create a ENSFactory like I did AsyncENSFactory so that there is consistency. The reason that AsyncENSFactory was created was to support the async creation of the web3 with the stalecheck middleware. If I create an ENSFactory then I would remove the init and fromweb3 from ENS which would be breaking but drive consistency between AsyncENS and ENS

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)
- [x] Need to add docs stalecheck and AsyncENS

### CODE REVIEW

File: ens/main.py
Comment (fselmo) [inline ens/main.py:82]: Thanks for the work @dbfreem! I haven't taken a look at any specifics yet but I would ideally like for this PR to be when we start to separate larger async and sync classes based on a similar structure to [this comment](https://github.com/ethereum/web3.py/pull/2441#pullrequestreview-966920702). Thoughts @kclowes?

I'd be happy to take that on if you'd like @dbfreem since you've thrown this all together already. It's late today but I plan on taking a closer look at specifics here this week. Just wanted to get the conversation started again on the sync + async organization moving forward.

Comment (dbfreem) [reply ens/main.py:82]: @fselmo I was thinking about doing that in the first place but then thought it might make the PR too big. I went ahead and reorganized it take a look and let me know what you think.

**Comment by fselmo** [comment]
hey @dbfreem... I had a chance to review most of this yesterday. I was on a plane so I just kind of made the changes locally and it was mostly cleanup except a few missing awaits so I just pushed it as a separate commit since the work was done already. Open for discussion on anything I moved around though.

My reasoning behind the nit / reorganizing the code style is since these are all new lines of code, it sets it up a bit nicer for cleaner git blame / commit history in the future. Other than that I just silenced some warnings that were already present in the previous code base that may have been copied over and added the two missing awaits. I also chose to cast to `Contract` and `AsyncContract` over using `# type: ignore`.

This is a great addition and a really nice example of how we should end up splitting our sync + async modules in the future so there were some pretty crucial convention choices solidified here 🎉 

I think it'd be nice to add the rest of the async tests such as for `test_offchain_lookup.py`... let me know if you're interested in tackling some of those. We are deploying those contracts on the async ENS setup so I think it'd be good to test it.

**Comment by fselmo** [comment]
I'd like to talk a bit about using the factory approach. I think the factory approach is a good way to handle the async init and I think I'm OK with the async setup being a bit different that sync since it was unavoidable really.

I don't mind _not_ adding an ENS factory for sync if it's not necessary... but curious to hear what you guys think @dbfreem, @kclowes, @pacrob. Thoughts?

**Comment by dbfreem** [comment]
> hey @dbfreem... I had a chance to review most of this yesterday. I was on a plane so I just kind of made the changes locally and it was mostly cleanup except a few missing awaits so I just pushed it as a separate commit since the work was done already. Open for discussion on anything I moved around though.
> 
> My reasoning behind the nit / reorganizing the code style is since these are all new lines of code, it sets it up a bit nicer for cleaner git blame / commit history in the future. Other than that I just silenced some warnings that were already present in the previous code base that may have been copied over and added the two missing awaits. I also chose to cast to `Contract` and `AsyncContract` over using `# type: ignore`.
> 
> This is a great addition and a really nice example of how we should end up splitting our sync + async modules in the future so there were some pretty crucial convention choices solidified here 🎉 
> 
> I think it'd be nice to add the rest of the async tests such as for `test_offchain_resolution.py`and `test_wildcard_resolution.py`... let me know if you're interested in tackling some of those. We are deploying those contracts on the async ENS setup so I think it'd be good to test it.

I totally welcome the collaboration so your changes are fine. I will take a look at adding those additional test.

**Comment by dbfreem** [comment]
@fselmo I added the test. It was a copy and paste of some of the existing test that were in those wildcard and offchain.

**Comment by dbfreem** [comment]
nevermind, still some work to do around the test.

**Comment by dbfreem** [comment]
ok @fselmo now this is ready. It was a lot of copy and paste of the existing test.

File: tests/ens/test_get_text.py
Comment (pacrob) [inline tests/ens/test_get_text.py:85]: accounts?

**Review by pacrob — APPROVED** [review]
one spelling nit, otherwise lgtm

File: ens/async_ens.py
Comment (fselmo) [inline ens/async_ens.py:121]: This was a bit to make these models homogeneous in the order of methods for better code readability:

i.e.
address, setup_address, name, setup_name ... etc ... _get_resolver, _set_resolver (internal methods near the end)

And line up ENS and AsyncENS to be ordered the same way.

**Comment by fselmo** [comment]
closed by #2526

**Review by fselmo — APPROVED** [review]
Hey @dbfreem, thanks so much for all your work on async. It has really helped get the ball rolling on so much of what was left.

I'm approving this here ~~but I just did an ultimate wrap up, added all the missing tests for async, and squashed your commits into one and both of mine into one with some good commit messages over at #2526.~~

Let me know if you have any remaining input. 

~~The reason for another PR again (as was the case with Async Contract) is mostly due to the merge commits. In the future, if a `git rebase upstream/master` is done on the branch and the differences are resolved (instead of `git merge upstream/master` and fixing the differences), this allows for us to squash commits much easier when cleaning up the PR since the commits don't have multiple parents as is sometimes the case.~~ (ultimately wasn't necessary, see edit)

Thanks again 👌

---

**edit:** In the end that wasn't necessary since I force-pushed my local cleaned up branch with the cherry-picked commits here which effectively got rid of the merge commits and was rebased against `master` locally... d'oh! Going to close that one instead and squash here so we can keep the conversation in one place.

PS: I'll squash once we get another approval so we can see the history a bit better.

**Comment by dbfreem** [comment]
I am all good.

**Comment by fselmo** [comment]
Hey @dbfreem, I played around with this a bit and was able to remove the `AsyncENSFactory`. I added some tests for the stalecheck middleware too and rebased with some of the formatting PRs that came through recently adding `black` to some of these files.

I put that into a separate PR (#2547) since that's a concrete divergence from this PR. I'm still going to do a bit of cleanup there but curious on any feedback. Anything we're missing?

**Comment by dbfreem** [comment]
> Hey @dbfreem, I played around with this a bit and was able to remove the `AsyncENSFactory`. I added some tests for the stalecheck middleware too and rebased with some of the formatting PRs that came through recently adding `black` to some of these files.
> 
> I put that into a separate PR (#2547) since that's a concrete divergence from this PR. I'm still going to do a bit of cleanup there but curious on any feedback. Anything we're missing?

That is awesome that you were able to remove the need for AsyncENSFactory. I have been busy lately and haven't had a chance to look over the full PR. Sorry.

**Comment by fselmo** [comment]
closed by #2547

**Pull request closed without merging**
